File: backend/app/schemas/task.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class TaskTemplateCreate(BaseModel):
    name: str
    instructions: str
    label_schema: dict


class TaskTemplateResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    instructions: str
    label_schema: dict
    version: int
    created_at: datetime

    model_config = {"from_attributes": True}


class TaskResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    example_id: uuid.UUID
    template_id: uuid.UUID
    status: str
    priority: int
    required_annotations: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TaskGenerateRequest(BaseModel):
    template_id: uuid.UUID
    dataset_id: uuid.UUID
    required_annotations: int = 2


class ModelSuggestionResponse(BaseModel):
    id: uuid.UUID
    task_id: uuid.UUID
    provider: str
    model_name: str
    suggestion: dict
    confidence: float | None
    raw_response: dict | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ModelSuggestionPayload(BaseModel):
    provider: str
    model_name: str
    score: float | None = None
    suggested_label: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AnnotationSummary(BaseModel):
    id: uuid.UUID
    label: dict
    confidence: int | None = None
    notes: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class TaskDetailResponse(TaskResponse):
    source_example_id: uuid.UUID
    dataset_id: uuid.UUID | None = None
    query: str = ""
    candidate_document: str = ""
    document_id: str | None = None
    example_metadata: dict = {}
    model_suggestion: ModelSuggestionPayload | None = None
    annotations: list[AnnotationSummary] = []

File: frontend/src/lib/schemas/task.ts
import { z } from "zod";

export const TaskResponseSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid(),
  example_id: z.string().uuid(),
  template_id: z.string().uuid(),
  status: z.enum(["created", "suggested", "assigned", "submitted", "needs_review", "resolved", "exported"]),
  priority: z.number(),
  required_annotations: z.number(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type TaskResponse = z.infer<typeof TaskResponseSchema>;

export const ModelSuggestionResponseSchema = z.object({
  id: z.string().uuid(),
  task_id: z.string().uuid(),
  provider: z.string(),
  model_name: z.string(),
  suggestion: z.record(z.unknown()),
  confidence: z.number().nullable(),
  raw_response: z.record(z.unknown()).nullable(),
  created_at: z.string(),
});

export type ModelSuggestionResponse = z.infer<typeof ModelSuggestionResponseSchema>;

export const TaskTemplateResponseSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid(),
  name: z.string(),
  instructions: z.string(),
  label_schema: z.record(z.unknown()),
  version: z.number(),
  created_at: z.string(),
});

export type TaskTemplateResponse = z.infer<typeof TaskTemplateResponseSchema>;
